File: Settembre/31.ArrayDiff-10Settembre.js
//todo Dati due array con valori stringa, restituisci un nuovo array contenente tutti i valori presenti in uno solo degli array.

//! L'array restituito deve essere ordinato alfabeticamente.

//? Tempo di esecuzione:
//* Non superato ma per poco

//?Codice scritto senza utilizzo di ChatGpt

function arrayDiff(arr1, arr2) {

  let finalArr = []

  if(arr1.length > arr2.length) {
  for(let i = 0; i < arr1.length; i++) if(!arr2.includes(arr1[i])) finalArr.push(arr1[i])
  }
  
  if(arr1.length < arr2.length) {
    for(let i = 0; i < arr2.length; i++) if(!arr1.includes(arr2[i])) finalArr.push(arr2[i])
  }

  if(arr1.length === arr2.length) {
  for(let i = 0; i < arr1.length; i++) {
  if(!arr2.includes(arr1[i]) || !arr1.includes(arr2[i])) {
    finalArr.push(arr1[i]);
    finalArr.push(arr2[i]);
  }}}

  let alfabFinalArr = finalArr.sort((a,b) => a.localeCompare(b))
  
  return alfabFinalArr;
}

console.log( arrayDiff(["two", "four", "five", "eight"], ["one", "two", "three", "four", "seven", "eight"]))

//?Codice scritto da ChatGpt

function arrayDiff(arr1, arr2) {
  const set1 = new Set(arr1);
  const set2 = new Set(arr2);

  return [...set1, ...set2]
    .filter(el => !(set1.has(el) && set2.has(el)))
    .sort((a, b) => a.localeCompare(b));
}


//? Versione Chat Gpt 

//* ✅ PRO

//* Hai capito il concetto di differenza simmetrica

//* Output ordinato correttamente

//* Funziona in alcuni casi semplici

//! ❌ CONTRO (importanti)

//! Logica troppo complessa: i if sui length non servono

//! Caso length === → bug: aggiungi elementi duplicati

//! includes dentro i loop → O(n²)

//! Codice difficile da mantenere

//todo Considerazioni: 

//! N.B.

File: Python/2.Settembre/31.Array_diff-10Settembre.py
# //Prompt ChatGpt: questo è il modello. Invierò consegna esercizio, codice e vorrei che tu mi dessi consigli

# //todo Consegna dell'esercizio

# Dati due array di stringhe, restituisci un nuovo array contenente tutti i valori presenti in un solo array.
# L'array restituito deve essere ordinato alfabeticamente.

# //todo


# //*Codice scritto senza utilizzo di ChatGpt

def array_diff(arr1, arr2):

    new_list = [];

    for word in arr1:
        if word in arr2:
            continue;
        else:
            new_list.append(word)

    for word in arr2:
        if word in arr1:
            continue;
        else:
            new_list.append(word)

    new_list.sort()
    return new_list

print(array_diff(["one", "two", "three", "four", "six"], ["one", "three", "eight"]))

# //*


# //* Codice scritto da ChatGpt

def array_diff(arr1, arr2):
    result = []

    for word in arr1:
        if word not in arr2:
            result.append(word)

    for word in arr2:
        if word not in arr1:
            result.append(word)

    return sorted(result)

# //*


# //* Pro e contro del mio codice

# //* Pro

# - La soluzione è corretta e restituisce il risultato richiesto.
# - Il codice è leggibile e facile da seguire.
# - L'ordinamento finale è gestito correttamente.
# - Hai utilizzato una logica chiara per confrontare i due array.

# //! Contro

# - Gli else sono superflui dopo un continue.
# - I punti e virgola non sono necessari in Python.
# - Con array molto grandi, i controlli con "in" diventano meno efficienti.
# - Il codice può essere leggermente semplificato.

# //*


# //* Consigli

# - Preferisci condizioni positive come "if word not in arr2".
# - Evita continue ed else quando non sono indispensabili.
# - Abituati a valutare la complessità temporale della soluzione.
# - Valuta l'utilizzo dei set quando non è necessario mantenere i duplicati.